### PR TITLE
Add support for bidirectional destroy dependencies

### PR DESCRIPTION
**Pull request by sebjacobs**
Prior to this commit if you defined a bidirectional relationship
between two models with destroy dependencies on both sides, a call to
`destroy` would result in an infinite callback loop.

Take the following relationship.

```
class Content < ActiveRecord::Base
  has_one :content_position, dependent: :destroy
end

class ContentPosition < ActiveRecord::Base
  belongs_to :content, dependent: :destroy
end
```

Calling `Content#destroy` or `ContentPosition#destroy` would result in
an infinite callback loop.

This commit changes the behaviour of `ActiveRecord::Callbacks#destroy`
so that it guards against subsequent callbacks.

Thanks to @zetter for demonstrating the issue with failing tests[1].

[1] rails#13609

### CODE REVIEW

File: activerecord/CHANGELOG.md [hunk 1]
```diff
@@ -328,4 +328,20 @@
...
+        class Content < ActiveRecord::Base
+          has_one :position, dependent: :destroy
+        end
+
+        class Position < ActiveRecord::Base
+          belongs_to :content, dependent: :destroy
+        end
+
+    *Seb Jacobs*
+
```
Comment (vipulnsward) [inline activerecord/CHANGELOG.md]: New CHANGELOG should be at the top of the file

Comment (sebjacobs) [inline activerecord/CHANGELOG.md]: @vipulnsward fixed in d5bf649.

**Comment by pixeltrix** [comment]
@sebjacobs I'm not sure whether this is a bug in how people are modelling their domain or whether it's a behaviour that needs to work. In both your example and the original issue's example the models shouldn't have the `:dependent` option on the `belongs_to` association. Why would you initiate the delete on the `Content::Position` model - it seems a contrived an example.

Similarly the example in the original issue has an address at the other end of a has many through which gets deleted when the link is deleted - this suggests that the address should be a direct has many to the user.

@rafaelfranca @sgrif wdyt?

**Comment by pixeltrix** [comment]
Ignoring the names for the moment and taking your associations, then it's straightforward enough to call `@position.content.destroy`. I think when deleting object graphs like this you need to have clear entry points for deletion - allowing cascading deletes from what is essentially a leaf node is asking for trouble.

Having said that we should definitely do better at handling it - maybe checking somehow when the association is built?

**Comment by sebjacobs** [comment]
@pixeltrix you are correct the example is rather contrived.

Perhaps the example below demonstrates a more realistic use case.

```ruby
class Course < ActiveRecord::Base
  has_many :steps
end

class Step < ActiveRecord::Base
  belongs_to :course
  belongs_to :content, polymorphic: true, dependent: :destroy
end

module Content
  has_one :step, as: :content, dependent: :destroy
end

class Article < ActiveRecord::Base; end
class VideoArticle < ActiveRecord::Base; end
class Quiz < ActiveRecord::Base; end
```

**Comment by sebjacobs** [comment]
> Having said that we should definitely do better at handling it - maybe checking somehow when the association is built?

It might also be worth documenting this issue.

**Comment by egilburg** [comment]
Does this problem only exist when both sides use `dependent: :destroy`, or also when one side uses `dependent: :destroy` and the other side has a manual destroy before/after hook?

Because I can think of the following example:

```ruby
class Post < ActiveRecord::Base
  has_many :comments, dependent: :destroy
end

class Comment < ActiveRecord::Base
  belongs_to :post
  after_destroy do
    post.destroy if post.comments.none?
  end
end
```

(The idea being, you are modelling a message board post+comment, where the post contains the overall thread subject, but the body of the post is saved as the first comment. Then, either the whole post can get deleted, or all the comments can get deleted, and if they do, the parent post should be deleted as well.

**Comment by daniel-rikowski** [comment]
Is there something I can do to help with this issue? With every new Rails version I have to pull the latest workaround from  https://github.com/rails/rails/issues/13609 :tired_face:

@egilburg Yes, the problem is not restricted to `dependent: :destroy`, but actually affects all `after_destroy` callbacks. (`dependent: :destroy` implicitly adds an `after_destroy` callback) Your code is a prime example for this problem, both callbacks will be called forever, unless one side stops calling `destroy`. (Which is what @sebjacobs code does, by checking if the callback has been called already)

The fact that everbody is talking about `dependent: :destroy` is a little bit misleading. @sebjacobs actually fixed a **circular callback problem**, which just often surfaces when using circular dependencies. 

So if this problem should be "fixed in the documentation" it should be done both in the callback and the association sections.

@pixeltrix 

>  I think when deleting object graphs like this you need to have clear entry points for deletion

I can understand your point of view but one could argue the same for circular auto-save associations (and their callbacks!), which are perfecly handled by Rails. 
In other words: Why should save callbacks have "loop protection" but destroy callbacks not? 

> Having said that we should definitely do better at handling it - maybe checking somehow when the association is built?

When first investigating this problem I tried to do that, but it didn't work because the problem is not limited to associations: Even if it were possible, it would still not help with manual `after_destroy` callbacks.

**Comment by Bertg** [comment]
We are currently in the works of upgrading a Rails 3.2 to Rails 4.1. We encountered this issue. 
Does anyone know if this has been dealt with in Rails 4.2?

**Comment by stevehanson** [comment]
@bertg Yes, I seem to be having the issue on Rails 4.2.3

**Comment by balexand** [comment]
This is great, thanks! :sparkles:

**Comment by arthurnn** [comment]
i am glad this was fixed. thanks all

**Comment by fernandomm** [comment]
Will this fix be applied in rails 4.2 or only in rails 5?

**Comment by NielsKSchjoedt** [comment]
I checked the code. It's only changed in rails 5
